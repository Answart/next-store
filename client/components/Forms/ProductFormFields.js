import PropTypes from 'prop-types';
import NProgress from 'nprogress';
import StyledProduct from '../styles/ProductStyles';
import ByCreator from '../ByCreator';
import { DEPARTMENTS, CATEGORIES } from '../../config';
import { capWord } from '../../lib/utils';
import { uploadImageFile, destroyImageFileByToken } from '../../lib/cloudinary';


const ProductFormFields = props => {
  const { department, image, online } = props;
  let categoriesByDept = department
    ? CATEGORIES[department]
    : CATEGORIES["tops"];
  const handleChange = e => {
    if (!!e.preventDefault) e.preventDefault();
    let { name, type, value, checked } = e.target;
    let state = {};
    let val = value;

    if (name === "department") state.category = "";
    if (type === "radio") {
      if (name === "online" || name === "offline") {
        val = (name === "online");
        name = "online";
      }
    }
    state[name] = val;

    props.saveToForm(state);
  };
  const handleImageChange = async e => {
    const { name, type, value, files } = e.target;
    const currentImageToken = props.image
      ? (props.image.delete_token || '')
      : '';
    NProgress.start();

    if (type === "file" && name === "image") {
      if (!files || !files.length) return;
      const image = await uploadImageFile(files[0]);
      if (image.error) return alert('An error occured while uploading image. Please try again later.');
      if (!!currentImageToken.length) await destroyImageFileByToken(currentImageToken);

      props.saveToForm({ image });
    }
    NProgress.done();
  }
  return (
    <StyledProduct>
      <div className="form-imgs">
        {!!image ? (
          <img width="450" height="640" src={image.large_image_url} alt={image.name} />
        ) : (
          <img width="450" height="640" src="/static/images/placeholder_large.jpg" alt="Placeholder Image" />
        )}
      </div>

      <div className="form-content">
        <div className="field-padding">
          <label htmlFor="title">
            Title:
            <input id="title"
              type="text"
              name="title"
              placeholder="UGG Classic Boot"
              value={props.title}
              onChange={handleChange}
              required
            />
          </label>
        </div>

        {props.userName && (
          <ByCreator
            name={props.userName}
            online={true}
          />
        )}

        <div className="field-padding">
          <label htmlFor="department">
            Department:
            <select id="department"
              name="department"
              value={department}
              onChange={handleChange}
              required
            >
              {DEPARTMENTS.map(dept => <option key={dept} value={dept}>{capWord(dept)}</option>)}
            </select>
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="category">
            Category:
            <select id="category"
              name="category"
              value={props.category}
              onChange={handleChange}
            >
              <option key={0} value=''></option>
              {categoriesByDept.map(ctgry => <option key={ctgry} value={ctgry}>{capWord(ctgry)}</option>)}
            </select>
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="description">
            Description:
            <textarea id="description"
              name="description"
              placeholder="UGG boots are a unisex style of sheepskin boot originating in Australia and New Zealand..."
              value={props.description}
              onChange={handleChange}
              required
            />
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="brand">
            Brand:
            <input id="brand"
              type="text"
              name="brand"
              placeholder="UGG"
              value={props.brand}
              onChange={handleChange}
            />
          </label>
        </div>

        <div className="field-padding">
          <label htmlFor="status">
            Status:
            <div className="prdct-padding">
              <div className="prdct-padding">
                <label htmlFor="offline" className="inline-lbl">
                  <input id="offline"
                    type="radio"
                    name="offline"
                    key={!online}
                    value={!online}
                    onChange={handleChange}
                    checked={!online}
                  />Private
                </label>
              </div>
              <div>
                <label htmlFor="online" className="inline-lbl">
                  <input id="online"
                    type="radio"
                    name="online"
                    key={online}
                    value={online}
                    onChange={handleChange}
                    checked={online}
                  />Public
                </label>
              </div>
            </div>
          </label>
        </div>

        <div className="field-padding">
          <label>
            Image:
            <div>
              <div className="prdct-padding">
                <label htmlFor="image" className="lbl-button">
                  <input id="image"
                    type="file"
                    name="image"
                    onChange={handleImageChange}
                  />Upload Image
                </label>
              </div>

              {image && !!image.name.length && (
                <div className="field-padding field-detail">
                  <h5>
                    Preview Details
                  </h5>
                  <p>
                    ID: <i>{image.cloudinary_id}</i>
                  </p>
                  <p>
                    Name: <i>{image.name}</i>
                  </p>
                  <p>
                    Dimensions: <i>{image.width}</i>W x <i>{image.height}</i>H
                  </p>
                  <p>
                    Transformation: <i>{image.transformation}</i>
                  </p>
                  <p>
                    Url: <i>{image.image_url}</i>
                  </p>
                </div>
              )}
            </div>
          </label>
        </div>
      </div>
    </StyledProduct>
  );
};

ProductFormFields.propTypes = {
  title: PropTypes.string.isRequired,
  userName: PropTypes.string,
  department: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  online: PropTypes.bool.isRequired,
  category: PropTypes.string,
  brand: PropTypes.string,
  image: PropTypes.shape({
    id: PropTypes.string,
    cloudinary_id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
    transformation: PropTypes.string.isRequired,
    image_url: PropTypes.string.isRequired,
    large_image_url: PropTypes.string.isRequired,
    delete_token: PropTypes.string
  }),
  saveToForm: PropTypes.func.isRequired
};


export default ProductFormFields;
