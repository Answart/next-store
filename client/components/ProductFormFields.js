import React, { Component } from 'react';
import PropTypes from 'prop-types';
import departments from '../lib/departments';
import categoriesByDept from '../lib/categoriesByDept';
import { user } from '../lib/dummyData';


class ProductFormFields extends Component {
  static propTypes = {
    title: PropTypes.string.isRequired,
    department: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    online: PropTypes.bool,
    image: PropTypes.string.isRequired,
    category: PropTypes.string,
    brand: PropTypes.string,
    handleChange: PropTypes.func.isRequired,
    uploadFile: PropTypes.func.isRequired,
    previewImage: PropTypes.bool
  };
  render() {
    const {
      title,
      department,
      description,
      online,
      image,
      category,
      brand,
      handleChange,
      uploadFile,
      previewImage
    } = this.props;
    let categories = department
      ? categoriesByDept[department]
      : categoriesByDept['Tops'];
    return (
      <div>
        {previewImage && (
          <div className="prdct-imgs">
            {image && (
              <img width="200" src={image} alt="Upload Preview" />
            )}
          </div>
        )}

        <div className="prdct-content">
          <label htmlFor="title" className="field-padding">
            Title:
            <input
              type="text"
              id="title"
              name="title"
              placeholder="Title"
              value={title}
              onChange={handleChange}
              required
            />
          </label>

          <div className="prdct-creator field-padding">
            By <strong>{user.name}</strong>
          </div>

          <label htmlFor="department">
            Department:
            <select
              id="department"
              name="department"
              form="carform"
              placeholder="Department i.e. Shoes"
              value={department}
              onChange={handleChange}
              required
            >
              {departments.map(dept => <option key={dept} value={dept}>{dept}</option>)}
            </select>
          </label>

          <label htmlFor="category">
            Category:
            <select
              id="category"
              name="category"
              form="carform"
              value={category}
              onChange={handleChange}
            >
              <option key={0} value={''}></option>
              {categories.map(ctgry => <option key={ctgry} value={ctgry}>{ctgry}</option>)}
            </select>
          </label>

          <label htmlFor="description">
            Description:
            <textarea
              id="description"
              name="description"
              placeholder="Enter A Description"
              value={description}
              onChange={handleChange}
              required
            />
          </label>

          <label htmlFor="brand">
            Brand:
            <input
              type="text"
              id="brand"
              name="brand"
              placeholder="Brand"
              value={brand}
              onChange={handleChange}
            />
          </label>

          <input
            type="checkbox"
            id="online"
            name="online"
            value={online}
            onChange={handleChange}
            checked={online ? "checked" : ""}
          />
          <label htmlFor="online" className="chkbx-label">
            Online
          </label>

          <label htmlFor="img">
            Image:
            <input
              type="file"
              id="img"
              name="img"
              placeholder="Upload an image"
              onChange={uploadFile}
              required
            />
          </label>
        </div>
      </div>
    );
  }
}

export default ProductFormFields;
