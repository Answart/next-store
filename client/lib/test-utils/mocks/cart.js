import {
  ADD_TO_CART_MUTATION,
  UPDATE_CARTITEM_MUTATION,
  REMOVE_FROM_CART_MUTATION
} from '../../../graphql'
import { fakeCartItem } from '../utils';
import { mockVariant } from './variant';

const mockCartItem = fakeCartItem();


const addToCartMutationMock = {
  request: { query: ADD_TO_CART_MUTATION, variables: { id: mockVariant.id } },
  result: {
    data: {
      addToCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};

const updateCartItemMutationMock = (quantity) => ({
  request: { query: UPDATE_CARTITEM_MUTATION, variables: { id: mockCartItem.id, quantity } },
  result: {
    data: {
      updateCartItem: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
        quantity,
        variant: { ...mockCartItem.variant }
      },
    },
  },
});

const removeFromCartMutationMock = {
  request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: mockCartItem.id } },
  result: {
    data: {
      removeFromCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};


export {
  mockCartItem,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
}
