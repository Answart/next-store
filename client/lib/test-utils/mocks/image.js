import { CREATE_IMAGE_MUTATION } from '../../../graphql'
import { fakeImage } from '../utils';

const mockImage = fakeImage();
const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
}


const createImageMutationMock = (overrides) => ({
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      ...overrides,
    },
  },
  result: {
    data: {
      createImage: {
        __typename: 'Image',
        id: mockImage.id,
      },
    },
  },
});

const createImageMutationErrorMock = {
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      cloudinary_id: '',
      name: '',
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};


export {
  mockImage, mockImageVariables,
  createImageMutationMock, createImageMutationErrorMock,
}
