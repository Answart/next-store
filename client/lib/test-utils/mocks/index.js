export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
