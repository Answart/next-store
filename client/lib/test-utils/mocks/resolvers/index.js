export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
} from './local';
export {
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
} from './user';
export {
  createImageMutationMock, createImageMutationErrorMock,
} from './image';
export {
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
} from './product';
export {
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock
} from './variant';
export {
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
} from './cartItem';
