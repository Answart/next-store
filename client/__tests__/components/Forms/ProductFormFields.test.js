import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import ProductFormFields from '../../../components/Forms/ProductFormFields';
import { mockUser, mockImage, mockProduct, createImageMutationErrorMock } from '../../../lib/test-utils/mocks';

const mocks = [
  { ...createImageMutationErrorMock },
];

describe('<ProductFormFields />', () => {
  let wrapper, saveToForm;
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider>
        <ProductFormFields mocks={mocks}
          title="" userName={mockUser.name} department="" description="" category="" brand="" online={false}
          image={{
            cloudinary_id: 'cloudinary_idcloudinary_id',
            name: 'namename',
            height: 33,
            width: 22,
            transformation: '',
            image_url: 'image_urlimage_url.png',
            large_image_url: 'large_image_urllarge_image_url.png',
            delete_token: 'delete tokeeeen',
          }}
          saveToForm={jest.fn()}
        />
      </MockedProvider>
    );
    saveToForm = wrapper.find('ProductFormFields').props().saveToForm;
  });
  afterAll(() => wrapper.unmount());
  afterEach(() => jest.clearAllMocks());

  it('renders and matches snapshot', async () => {
    const form = wrapper.find('ProductFormFields');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it('renders the ByCreator component', async () => {
    const byCreator = wrapper.find('ByCreator');
    expect(byCreator.text()).toEqual(`By${mockUser.name}`);
    expect(byCreator.props().name).toEqual(mockUser.name);
    expect(byCreator.props().online).toEqual(true);
  });

  describe('input changes', async () => {

    it('title input works', async () => {
      wrapper.find('#title').simulate('change', { target: {
        value: mockProduct.title, name: 'title',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ title: mockProduct.title });
    });

    it('department select input works', async () => {
      wrapper.find('#department').simulate('change', { target: {
        value: mockProduct.department, name: 'department', type: 'select',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ department: mockProduct.department, category: '' });
    });

    it('category select input works', async () => {
      wrapper.find('#category').simulate('change', { target: {
        value: mockProduct.category, name: 'category', type: 'select',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ category: mockProduct.category });
    });

    it('description input works', async () => {
      wrapper.find('#description').simulate('change', { target: {
        value: mockProduct.description, name: 'description',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ description: mockProduct.description });
    });

    it('brand input works', async () => {
      wrapper.find('#brand').simulate('change', { target: {
        value: mockProduct.brand, name: 'brand',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ brand: mockProduct.brand });
    });

    it('status inputs online/offline radio button work', async () => {
      wrapper.find('#online').simulate('change', { target: {
        value: mockProduct.online, name: 'online', type: 'radio',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ online: !mockProduct.online });
      wrapper.find('#offline').simulate('change', { target: {
        value: mockProduct.online, name: 'offline', type: 'radio',
      }});
      wrapper.update();
      expect(saveToForm).toHaveBeenCalled();
      expect(saveToForm).toHaveBeenCalledWith({ online: mockProduct.online });
    });

    describe('image/cloudinary upload input', async () => {
      afterEach(() => global.fetch.mockReset())

      it('fails silently if input empty', async () => {
        global.fetch = jest.fn().mockResolvedValue({
          json: () => ({})
        });
        wrapper.find('input[type="file"]').simulate('change', { target: {
          files: [], name: 'image', type: 'file',
        }});
        await wait(0);
        expect(global.fetch).not.toHaveBeenCalled();
        expect(saveToForm).not.toHaveBeenCalled();
      });

      it('returns alert message when cloudinary upload fails', async () => {
        console.error = jest.fn();
        // mock the global fetch API (invalid res cloudinary)
        global.fetch = jest.fn().mockResolvedValue({
          json: () => ({
            public_id: mockImage.cloudinary_id,
            original_filename: mockImage.name,
            secure_url: mockImage.image_url,
          }),
        });
        wrapper.find('input[type="file"]').simulate('change', { target: {
          files: [mockImage.image_url], name: 'image', type: 'file',
        }});
        await wait(0);
        expect(global.fetch).toHaveBeenCalled();
        expect(console.error).toHaveBeenCalled();
        expect(console.error).toHaveBeenCalledWith("Error creating image file in cloudinary uploadImageFile. TypeError: Cannot read property 'length' of undefined");
        expect(window.alert).toHaveBeenCalled();
        expect(window.alert).toHaveBeenCalledWith('An error occured while uploading image. Please try again later.');
        expect(saveToForm).not.toHaveBeenCalled();
      });

      describe('successful file upload', async () => {
        beforeEach(() => {
          // mock the global fetch API (cloudinary)
          global.fetch = jest.fn().mockResolvedValue({
            json: () => ({
              public_id: mockImage.cloudinary_id,
              original_filename: mockImage.name,
              secure_url: mockImage.image_url,
              delete_token: 'delete tokeeeen',
              eager: [{
                height: mockImage.height,
                width: mockImage.width,
                transformation: mockImage.transformation,
                secure_url: mockImage.large_image_url,
              }]
            }),
          });
          wrapper.find('input[type="file"]').simulate('change', { target: {
            files: [mockImage.image_url], name: 'image', type: 'file',
          }});
        });

        it('uploads file to cloudinary and saves returned image to form', async () => {
          await wait(0);
          expect(global.fetch).toHaveBeenCalled();
          expect(saveToForm).toHaveBeenCalled();
        });

        it('calls cloudinary to delete previous image if replacing an already uploaded image', async () => {
          await wait(0);
          wrapper.find('input[type="file"]').simulate('change', { target: {
            files: ['anotherfile.png'], name: 'image', type: 'file',
          }});
          await wait(0);
          expect(global.fetch).toHaveBeenCalled();
          expect(saveToForm).toHaveBeenCalled();
        });
      });
    })
  });
});
