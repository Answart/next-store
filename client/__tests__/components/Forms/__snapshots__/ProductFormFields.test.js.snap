// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductFormFields /> renders and matches snapshot 1`] = `
<ProductFormFields
  brand=""
  category=""
  department=""
  description=""
  image={
    Object {
      "cloudinary_id": "cloudinary_idcloudinary_id",
      "delete_token": "delete tokeeeen",
      "height": 33,
      "image_url": "image_urlimage_url.png",
      "large_image_url": "large_image_urllarge_image_url.png",
      "name": "namename",
      "transformation": "",
      "width": 22,
    }
  }
  mocks={
    Array [
      Object {
        "request": Object {
          "query": Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "CREATE_IMAGE_MUTATION",
                },
                "operation": "mutation",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "cloudinary_id",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "cloudinary_id",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "name",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "name",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "height",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "height",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "width",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "width",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "transformation",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "transformation",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "image_url",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "image_url",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "large_image_url",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "large_image_url",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "createImage",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "cloudinary_id",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "Int",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "height",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "Int",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "width",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "transformation",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "image_url",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "directives": Array [],
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "String",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "large_image_url",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 464,
              "start": 0,
            },
          },
          "variables": Object {
            "cloudinary_id": "",
            "height": 22,
            "image_url": "peggswatch1.jpg",
            "large_image_url": "peggswatch2.jpg",
            "name": "",
            "transformation": "",
            "width": 22,
          },
        },
        "result": Object {
          "errors": Array [
            Object {
              "message": "ack!",
            },
          ],
        },
      },
    ]
  }
  online={false}
  saveToForm={[MockFunction]}
  title=""
  userName="Miss Coleman Berge"
>
  <ProductStyles__StyledProduct>
    <div
      className="ProductStyles__StyledProduct-sc-2231u4-0 dOvnhj"
    >
      <div
        className="form-imgs"
      >
        <img
          alt="namename"
          height="640"
          src="large_image_urllarge_image_url.png"
          width="450"
        />
      </div>
      <div
        className="form-content"
      >
        <div
          className="field-padding"
        >
          <label
            htmlFor="title"
          >
            Title:
            <input
              id="title"
              name="title"
              onChange={[Function]}
              placeholder="UGG Classic Boot"
              required={true}
              type="text"
              value=""
            />
          </label>
        </div>
        <ByCreator
          name="Miss Coleman Berge"
          online={true}
        >
          <div
            className="prdct-creator prdct-padding"
          >
            By
            <Link
              href={
                Object {
                  "pathname": "/shop",
                  "query": Object {
                    "name": "Miss Coleman Berge",
                    "online": true,
                  },
                }
              }
            >
              <a
                className="undrln-btn"
                href="/shop?name=Miss%20Coleman%20Berge&online=true"
                onClick={[Function]}
              >
                Miss Coleman Berge
              </a>
            </Link>
          </div>
        </ByCreator>
        <div
          className="field-padding"
        >
          <label
            htmlFor="department"
          >
            Department:
            <select
              id="department"
              name="department"
              onChange={[Function]}
              required={true}
              value=""
            >
              <option
                key="tops"
                value="tops"
              >
                Tops
              </option>
              <option
                key="bottoms"
                value="bottoms"
              >
                Bottoms
              </option>
              <option
                key="shoes"
                value="shoes"
              >
                Shoes
              </option>
              <option
                key="outwear"
                value="outwear"
              >
                Outwear
              </option>
              <option
                key="accessories"
                value="accessories"
              >
                Accessories
              </option>
              <option
                key="decor"
                value="decor"
              >
                Decor
              </option>
              <option
                key="wedding"
                value="wedding"
              >
                Wedding
              </option>
            </select>
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="category"
          >
            Category:
            <select
              id="category"
              name="category"
              onChange={[Function]}
              value=""
            >
              <option
                key="0"
                value=""
              />
              <option
                key="blouses"
                value="blouses"
              >
                Blouses
              </option>
              <option
                key="cartigans"
                value="cartigans"
              >
                Cartigans
              </option>
              <option
                key="casual"
                value="casual"
              >
                Casual
              </option>
              <option
                key="sweaters"
                value="sweaters"
              >
                Sweaters
              </option>
              <option
                key="sweatshirts"
                value="sweatshirts"
              >
                Sweatshirts
              </option>
            </select>
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="description"
          >
            Description:
            <textarea
              id="description"
              name="description"
              onChange={[Function]}
              placeholder="UGG boots are a unisex style of sheepskin boot originating in Australia and New Zealand..."
              required={true}
              value=""
            />
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="brand"
          >
            Brand:
            <input
              id="brand"
              name="brand"
              onChange={[Function]}
              placeholder="UGG"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label
            htmlFor="status"
          >
            Status:
            <div
              className="prdct-padding"
            >
              <div
                className="prdct-padding"
              >
                <label
                  className="inline-lbl"
                  htmlFor="offline"
                >
                  <input
                    checked={true}
                    id="offline"
                    key="true"
                    name="offline"
                    onChange={[Function]}
                    type="radio"
                    value={true}
                  />
                  Private
                </label>
              </div>
              <div>
                <label
                  className="inline-lbl"
                  htmlFor="online"
                >
                  <input
                    checked={false}
                    id="online"
                    key="false"
                    name="online"
                    onChange={[Function]}
                    type="radio"
                    value={false}
                  />
                  Public
                </label>
              </div>
            </div>
          </label>
        </div>
        <div
          className="field-padding"
        >
          <label>
            Image:
            <div>
              <div
                className="prdct-padding"
              >
                <label
                  className="lbl-button"
                  htmlFor="image"
                >
                  <input
                    id="image"
                    name="image"
                    onChange={[Function]}
                    type="file"
                  />
                  Upload Image
                </label>
              </div>
              <div
                className="field-padding field-detail"
              >
                <h5>
                  Preview Details
                </h5>
                <p>
                  ID: 
                  <i>
                    cloudinary_idcloudinary_id
                  </i>
                </p>
                <p>
                  Name: 
                  <i>
                    namename
                  </i>
                </p>
                <p>
                  Dimensions: 
                  <i>
                    22
                  </i>
                  W x 
                  <i>
                    33
                  </i>
                  H
                </p>
                <p>
                  Transformation: 
                  <i />
                </p>
                <p>
                  Url: 
                  <i>
                    image_urlimage_url.png
                  </i>
                </p>
              </div>
            </div>
          </label>
        </div>
      </div>
    </div>
  </ProductStyles__StyledProduct>
</ProductFormFields>
`;
